### DISCUSSION TITLE
Error in project 2 when running functions

### DISCUSSION THREAD

**Discussion by 2199105T** [IBAHCM/RPiR#336] [Archive] [answered]
Hi @richardreeve @soniamitchell ,

In practical 2, I am trying to write the new species richness and Simpson diversity functions using the general diversity function I wrote but every time I go to test them with a population I get this error message:

```{r}
Error: C stack usage  7970112 is too close to the limit
```

**Comment by soniamitchell** [comment, marked as answer]
Hey @2199105T, in your functions you've used `UseMethod()`, which is a way of defining methods in object oriented style programming (which we're not asking you to try). You've some how created a recursive loop and that's maxed out your memory. If you remove `UseMethod()` you should be fine 👍

**Reply by 2199105T** [reply]
so just put...

p2_general_diversity(q==2)?

**Reply by soniamitchell** [reply]
If the first argument of your `p2_general_diversity()` function requires `TRUE` or `FALSE` that would be fine, because `q==0` will return `TRUE` or  `FALSE` depending on the value of  `q`.

**Reply by 2199105T** [reply]
I removed UseMethod and now I get this error:

```{r}
Error in p2_general_diversity(q == 2) : 
  argument "q" is missing, with no default
```

**Reply by soniamitchell** [reply]
That suggests a problem with your function, specifically that you're missing an argument, `q`. 

Sorry, I can't give you specific guidance because this is your final project.

**Reply by 2199105T** [reply]
that's okay, can you tell me if you mean my general diversity function or my specific functions?

**Reply by soniamitchell** [reply]
This error message:

> Error in p2_general_diversity(q == 2) : 
  argument "q" is missing, with no default

is telling you that there's an error in the call to `p2_general_diversity(q == 2)`, specifically that the argument `q` is missing.

**Reply by richardreeve** [reply]
Hint: `q==2` tests whether a variable you already have called `q` is equal to 2, whereas `q=2` sets an argument called `q` in a function you are calling to 2...
